modtools/fire-rate.lua

```diff
@@ -0,0 +1,511 @@
+:-fortCap <integer> -advCap <integer>:
+  Sets a cap on the fastest fire rate that can be achieved in their respective
+  mode. Due to game limitations, the cap can't be less than 10 in adventure
+  mode. Defaults to half of the base fire rate defined by the ``-fort`` or
+  ``-adv`` arguments.
+
+Other:
+
+:-mode <"basic" | "vanilla">:
+  Sets what method is used to determine how skill affects fire rates. This is
+  applied globally, rather than on a per-item basis. Basic uses a simplified
+  method for working out fire rates - each point in a skill reduces the fire
+  cooldown by a consistent, fixed amount. This method is the default.
+  Vanilla mode attempts to replicate behaviour for fire rates - skill rolls
+  determine which of 6 fixed increments of speeds is used, with a unit's skill
+  affecting the range and averages. **NOT YET IMPLEMENTED!**
+
+]====]
+
+--[[ Here's my current research on how I think this stuff works (v 47.05)
+Fort Mode uses a unit's think_counter. After firing a shot, it's set to 80 (not
+accounting for skill stuff). This delays them from acting again until then.
+
+If you make it so onProjItemCheckMovement sets their think_counter to 0 whenever
+they fire something, they can fire 1 shot every tick.
+
+Adventure mode uses think_counter + a fixed delay. After firing a shot, the game
+advances by the fixed delay, then starts decrementing the think_counter. The
+fixed delay is either 10 or 9 from testing. It probably exists to prevent units
+walking into their own shots.
+
+Example:
+- Throw - think_counter 1, total time passed 10
+- Spit - think_counter 1, total time passed 10
+
+When in tactical mode, the fixed delay always occurs after making a ranged
+attack before switching control to the next character.
+
+Then, the think counter delay prevents switching back into that character until
+it's passed.
+
+Some figures might be off by 1 because of delays between firing and when
+onProjItemCheckMovement triggers.
+]]
+
+---------------------------------------------------------------------
+local utils = require("utils")
+local eventful = require("plugins.eventful")
+
+local validArgs = utils.invert({
+  "help",
+  "item",
+  "throw",
+  "material",
+  "fortBase",
+  "advBase",
+  "advSkillFactor",
+  "fortSkillFactor",
+  "fortCap",
+  "advCap",
+  "mode",
+})
+
+-- The (hopefully unique) separator used to split a weapon token and material
+-- token when they're combined to make a key
+local separator = "|||"
+
+-- The (hopefully unique) key used as a substitute for a weapon token when the
+-- attack is coming from a throw.
+local throw_key = "throw"
+
+---------------------------------------------------------------------
+weapon_data = weapon_data or {}
+config = config or {
+  skill_mode = "basic",
+}
+---------------------------------------------------------------------
+
+function register_weapon(weapon_token, material_token, data)
+  local key = weapon_token
+
+  if ( material_token ~= nil ) then
```

Comment:
in lua, the preferred form would be:
`if not material_token then`
that is, just test the var for "truthiness" and don't use parentheses unless you have to for grouping

Reply:
Addressing a few of the formatting points: I tend to try and make things as clear as possible (even if it means slightly longer code), because I know that one of the main ways for beginners getting in to making DFhack scripts is to read existing ones to learn from.